pkg/scheduler/scheduler.go

```diff
@@ -54,9 +53,11 @@ const (
 )
 
 const (
-	scheduleSuccessReason  = "BindingScheduled"
-	scheduleFailedReason   = "BindingFailedScheduling"
-	scheduleSuccessMessage = "Binding has been scheduled"
+	scheduleSuccessReason    = "BindingScheduled"
+	scheduleFailedReason     = "BindingFailedScheduling"
+	noClusterAvailableReason = "NoClusterAvailable"
```

Comment:
This two types is confusing, what's the difference?

I'd like to refer to what kubernetes does.

1. We can't find any clusters that pass the filter plugins.
```
    message: 0/1 clusters are available: 1 cluster(s) didn't match the placement cluster affinity constraint.
    reason: Unschedulable
    status: "False"
    type: BindingScheduled
```

2. we can't assign replicas due to not enough resources.

```
    message: failed to assignReplicas: xxxxxx
    reason: Unschedulable
    status: "False"
    type: BindingScheduled
```

This two types will cover the unschedulable condition.

Reply:
`NoCluster` is for the scenarios without clusters. As discussed before, we need to distinguish between scenarios with and without clusters. For scenarios without clusters, retry scheduling actually does not make any sense, but from our requirements, we need to regard it as a successful scheduling. If we introduce `Unschedulable`, how to define it?

Reply:
Can you enlighten me about what is `without clusters`? Do you mean no clusters pass the filter plugin?

Reply:
yeah, and no member cluster

Reply:
>  For scenarios without clusters, retry scheduling actually does not make any sense

What if a new cluster is added, or a taint is removed, should we retry?

I don't think this is a successful scheduling. For kubernetes, a pod is not scheduled successfully unless it is truely assigned to a node. We can easily distinguish no-cluster availablity and scheduler internal error by `reason`. `Unschedulable` is a typical reason for no more clusters fit or no more enough resources.

Reply:
cc @XiShanYongYe-Chang @chaunceyjiang for opinions.

Reply:
> What if a new cluster is added, or a taint is removed, should we retry?

yeah, will retry scheduling driven by cluster events.

> Unschedulable is a typical reason for no more clusters fit or no more enough resources.

Sure, no more enough resources is failed scheduling, it is essentially different from the reason for no cluster. Just like insufficient CPU is also no more enough resources, however CPU resources are dynamically changing. When CPU resources return to normal, the scheduling can be successfully by retrying the scheduling. But in a scenario without any clusters, it will never be possible to heal itself without human intervention.

Reply:
I only see two kind of errors for NoClusterError.

1. No member clusters.
2. No fit clusters.

Actually `No member clusters` is a special kind of `No fit clusters`. So basically they belong together. I guess you are concerned about the retrying because we don't need to retry when there are no member clusters. This issue should not confuse with conditions and I think this is the duty for **scheduling queue**.

Now let's figure out what **successful scheduling** means. In analogy to this definition for Kubernetes, I could give a definition like this:

The binding.Spec.Clusters are updated by scheduler when a binding is passed all steps of scheduler. Otherwise it fails and we do have different kind of reasons or conditions. **Correct me if I'm wrong here.**

1. `BindingReasonUnschedulable = "SchedulerError"`, means that some internal error happens during scheduling, for example due to nodeAffinity parsing errors.
2. `PodReasonUnschedulable = "Unschedulable"`, means that the scheduler can't schedule the binding right now, for example due to no candidate cluters or insufficient resources in candidate clusters.

This two binding scheduling failure reasons are the same as kube-scheduler. So I don't know if there are more reasons here or do you want to seperate this `Unschedulable` reason with more precise ones?

Reply:
I remember the conclusion of our previous community meeting was to add new reason to distinguish between errors during scheduling and no suitable cluster after scheduling. Hi @Garrybest, do you mean to use `BindingFailedSchedulingreason` to indicate all scheduling failures, and then use meassage to distinguish the above situations?

Reply:
Emmm, not really. `BindingScheduled` is the type. And `SchedulerError`, `Unschedulable` is the reasons. Note that we distinguish failure and success by `status`. 

1. Success.
```
    status: "True"
    type: BindingScheduled
```
2. Internal Error
```
    message: internal error: xxxx
    reason: SchedulerError
    status: "False"
    type: BindingScheduled
```
3. Unschedulable
```
    message: 0/1 clusters are available: 1 cluster(s) didn't match the placement cluster affinity constraint.
    reason: Unschedulable
    status: "False"
    type: BindingScheduled
```

Reply:
@Garrybest thanks,  it's a good suggestion, but I really want to distinguish between the internal error and the error for no suitable cluster. If it is designed like this, I can only distinguish it by message, which will be very bad.

Reply:
Got it. But I guess `no suitable cluster` is a Unschedulable one and `internal error` is a `SchedulerError` one, isn't it?

Reply:
of course you are right. can we seperate the `Unschedulable` reason into more precise ones? maybe we don't need `Unschedulable` but seperate it into specific reasons, like `NoClusterAvailable`, `NoClusterFit` and `InsufficientResources` etc. And the internal error is still `SchedulerError`. how do you feel about this?

Reply:
I'm ok with it. But I'm not sure the difference between NoClusterAvailable and NoClusterFit. Can we merge them?

Reply:
sure, we can name it `NoClusterFit`. but for `NoClusterFit`, I think we do not need to return error to retry scheduling, cause it makes no sense and resources will be scheduled when new member cluster is joined or scheduling requirements are met.
Let's summarize, it looks like this:
1.  Unschedulable: no cluster fit
```
message: no clusters available to schedule
reason: NoClusterFit
status: "False"
type: Scheduled
```

2.  Unschedulable: insufficient resources 
```
message: cluster xxx has no sufficient memory
reason: InsufficientResources
status: "False"
type: Scheduled
```

3.  Internal error: eg. can not access api-server...
```
message: xxx timeout
reason: SchedulerError
status: "False"
type: Scheduled
```

4.  schedule successfully
```
message: Binding has been scheduled
reason: BindingScheduled
status: "True"
type: Scheduled
```

Reply:
Great! Now we have reached an agreement.